File: 725-split-linked-list-in-parts/split-linked-list-in-parts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        ans = [None] * k

        size = 0

        curr = head
        while curr:
            size += 1
            curr = curr.next

        splitLength = size // k
        r = size % k

        curr = head

        for i in range(k):
            dummyHead = ListNode(0)
            tail = dummyHead

            currSplit = splitLength
            if r > 0:
                r-= 1
                currSplit += 1
            
            for j in range(currSplit):
                tail.next = ListNode(curr.val)
                tail = tail.next
                curr = curr.next
            ans[i] = dummyHead.next
        

        return ans

File: 725-split-linked-list-in-parts/split-linked-list-in-parts.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function splitListToParts(head: ListNode | null, k: number): Array<ListNode | null> {
    // initial array with k size fill with null
    let ans = new Array(k).fill(null)

    // find the size of the listNode
    let size = 0
    let curr = head

    while(curr) {
        size++
        curr = curr.next
    }

    // find split length and remainder (if theres any)
    let splitLength = Math.floor(size / k)
    let remainder = size % k

    curr = head
    let prev = curr
    // loop through k
    for (let i = 0; i < k; i++) {
        let newPart = curr

        let currSize = splitLength
        if(remainder > 0) {
            currSize++
            remainder--
        }
        for (let j = 0; j < currSize; j++) {
            prev = curr
            if (curr !== null) {
                curr = curr.next
            }
        }
        if (prev !== null) {
            prev.next = null
        }
        ans[i] = newPart
    }

    return ans
};